spec/std/http/server/handlers/log_handler_spec.cr

```diff
@@ -1,35 +1,57 @@
 require "spec"
 require "http/server/handler"
+require "../../../../support/log"
+require "../../../../support/io"
 
 describe HTTP::LogHandler do
   it "logs" do
     io = IO::Memory.new
     request = HTTP::Request.new("GET", "/")
+    request.remote_address = "192.168.0.1"
     response = HTTP::Server::Response.new(io)
     context = HTTP::Server::Context.new(request, response)
 
     called = false
-    log_io = IO::Memory.new
-    handler = HTTP::LogHandler.new(log_io)
+    handler = HTTP::LogHandler.new
     handler.next = ->(ctx : HTTP::Server::Context) { called = true }
-    handler.call(context)
-    log_io.to_s.should match %r(GET / - 200 \(\d+(\.\d+)?[mµn]s\))
+    logs = capture_logs("http.server") { handler.call(context) }
+    match_logs(logs,
+      {:info, %r(^192.168.0.1 - GET / HTTP/1.1 - 200 \(\d+(\.\d+)?[mµn]s\)$)}
+    )
     called.should be_true
   end
 
-  it "does log errors" do
+  it "logs to custom logger" do
+    request = HTTP::Request.new("GET", "/")
+    response = HTTP::Server::Response.new(IO::Memory.new)
+    context = HTTP::Server::Context.new(request, response)
+
+    backend = Log::MemoryBackend.new
+    log = Log.new("custom", backend, :info)
+    handler = HTTP::LogHandler.new(log)
+    handler.next = ->(ctx : HTTP::Server::Context) {}
+    handler.call(context)
+
+    match_logs(backend.entries,
+      {:info, %r(^- - GET / HTTP/1.1 - 200 \(\d+(\.\d+)?[mµn]s\)$)}
+    )
+  end
+
+  it "log failed request" do
     io = IO::Memory.new
     request = HTTP::Request.new("GET", "/")
     response = HTTP::Server::Response.new(io)
     context = HTTP::Server::Context.new(request, response)
 
-    called = false
-    log_io = IO::Memory.new
-    handler = HTTP::LogHandler.new(log_io)
+    handler = HTTP::LogHandler.new
     handler.next = ->(ctx : HTTP::Server::Context) { raise "foo" }
-    expect_raises(Exception, "foo") do
-      handler.call(context)
+    logs = capture_logs("http.server") do
+      expect_raises(Exception, "foo") do
+        handler.call(context)
+      end
     end
-    (log_io.to_s =~ %r(GET / - Unhandled exception:)).should be_truthy
+    match_logs(logs,
+      {:info, %r(^- - GET / HTTP/1.1 - 200 \(\d+(\.\d+)?[mµn]s\)$)}
```

Comment:
The logged status code is plainly wrong here. The server actually responds with 500, not 200 and the log must reflect that.
I'm not sure what's the best way to make this accurate. The `LogHandler` probably needs to differentiate whether the handler chain was successfully completed or an exception was raised. The exception part would then need to duplicate some of the logic from `RequestProcessor` to tell the expected response status.

Another option would be to integrate `LogHandler` directly into `RequestProcessor`. This would avoid duplication of the error handling and centralize logging to a single place. It would obviously need to be configurable.

The purpose of `ErrorHandler` should also be considered in this regard. It seems quite useless when its only effect is to change this:
```
I, [2020-04-22T15:21:41.548787000Z #5279]   INFO -- crystal-run-server.tmp:http.server: 127.0.0.1:38650 - GET / HTTP/1.1 - 200 (74.30µs)
E, [2020-04-22T15:21:41.548973000Z #5279]  ERROR -- crystal-run-server.tmp:http.server: Unhandled exception on HTTP::Handler -- Exception: foo
```
To this:
```
E, [2020-04-22T15:31:56.224456000Z #5412]  ERROR -- crystal-run-server.tmp:http.server: Unhandled exception -- Exception: foo
I, [2020-04-22T15:31:56.224721000Z #5412]   INFO -- crystal-run-server.tmp:http.server: 127.0.0.1:50044 - GET / HTTP/1.1 - 500 (376.90µs)
```

Reply:
I also realized the problem with the status codes. That happens when only `LogHandler` is used. In Rack the log handler writes logs without handling errors so when this happens the request is not even logged. I think it's better (not ideal though) to log the wrong status code than not logging anything at all.

The purpose of `ErrorHandler` is handling errors, showing them to the client and writing the logs. I think the issue is the placement of the `LogHandler` and the fact that both handlers must be used at the same time to work properly. Rack also has the same "problem" but it setups both handlers automatically by default. Maybe we could do that but probably this could be better if the design is changed a little bit. Maybe `LogHandler` cannot be a regular `HTTP::Handler` and it should be a strategy instead, plugged in the `RequestProcessor`.